src/main/groovy/life/qbic/portal/sampletracking/web/views/SampleImport.groovy

```diff
@@ -57,25 +62,39 @@ class SampleImport extends VerticalLayout {
      */
     private def selectSampleFromTextfield() {
         // Get value from user input in textField
-        String sampleIdInput = additionalSampleId.getValue()
-        try {
-            if (sampleIdInput?.trim()) {
-
-                controller.selectSampleById(sampleIdInput)
-                // if sample was found show success notification
-                // TODO: Remove this and add a listener to the notification observable list
-                StyledNotification uploadIdSuccessNotification = new StyledNotification("Success", "Added $sampleIdInput")
-                uploadIdSuccessNotification.show(Page.getCurrent())
+        String sampleId = additionalSampleId.getValue()
+        boolean isSampleSelected = viewModel.samples.any { sample -> (sample as Sample)?.getCode() == sampleId }
 
+        try {
+            if (sampleId?.trim()) {
+                if (isSampleSelected) {
+                    viewModel.failureNotifications.add("You already selected ${sampleId}")
+                    log.warn("Tried to select ${sampleId} multiple times.")
+                } else {
+                    controller.selectSampleById(sampleId)
+                }
             } else {
-                StyledNotification emptyId = new StyledNotification("Empty ID", "Please enter a sample ID and try again.", Notification.Type.ASSISTIVE_NOTIFICATION)
-                emptyId.show(Page.getCurrent())
-
+                viewModel.failureNotifications.add("Please enter a sample ID and try again.")
             }
         } catch (Exception e) {
-            log.error("Unexpected error trying to add sample by id $sampleIdInput", e)
-            StyledNotification couldNotSelectSampleNotification = new StyledNotification("Could not select sample $sampleIdInput", Notification.Type.ERROR_MESSAGE)
-            couldNotSelectSampleNotification.show(Page.getCurrent())
+            log.error("Unexpected error trying to add sample by id $sampleId", e)
+            viewModel.failureNotifications.add("Could not select sample $sampleId")
+        }
+    }
+
+    private selectSamplesFromFile(List<String> sampleIds) {
+        List<String> alreadySelectedIds = viewModel.samples.collect { sample -> (sample as Sample).getCode() }
+        List<String> unselectedIds = sampleIds.findAll { sampleId -> !(sampleId in alreadySelectedIds) }
+        if (unselectedIds.size() < 1) {
+            log.debug("No new sample codes in list.")
+            viewModel.failureNotifications.add("There are no new sample codes in the selected file.")
```

Comment:
I would keep this in. In case the user uploads the file twice he/she will quickly find the mistake him/herself and won't have to wonder why nothing is changing. Should not be triggered normally but does not hurt I guess.

Reply:
Yes, but we should avoid implementations that were not specified by the users :)

Reply:
I guess @wow-such-code has to decide that